<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>How to Use &mdash; GRAPE 1.7 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=392da36c"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Dockstore" href="dockstore.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            GRAPE
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">How to Use</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#quality-control-and-data-preprocessing">Quality Control and Data Preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#preprocessing-steps">Preprocessing steps</a></li>
<li class="toctree-l3"><a class="reference internal" href="#workflows-usage">Workflows Usage</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#relatedness-inference">Relatedness Inference</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id1">Workflows Usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#description-of-the-ibis-and-ersa-parameters">Description of the IBIS and ERSA Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#description-of-the-output-relatives-file">Description of the Output Relatives File</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ibd-segments-weighting">IBD Segments Weighting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome-regions-exclusion-mask">Genome Regions Exclusion Mask</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ibd-segments-weighing">IBD Segments Weighing</a></li>
<li class="toctree-l3"><a class="reference internal" href="#computation-of-the-ibd-segments-weighing-mask">Computation of the IBD segments weighing mask</a></li>
<li class="toctree-l3"><a class="reference internal" href="#usage-of-the-ibd-segments-weighing-mask">Usage of the IBD segments weighing mask</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#execution-by-scheduler">Execution by Scheduler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dockstore.html">Dockstore</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GRAPE</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">How to Use</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/howto.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="how-to-use">
<span id="sec-how-to-use"></span><h1>How to Use<a class="headerlink" href="#how-to-use" title="Link to this heading"></a></h1>
<p>GRAPE requires a single multi-sample VCF file as input and has a separate workflow for downloading and checking the integrity of reference datasets.</p>
<p>The pipeline has three main steps:</p>
<ol class="arabic simple">
<li><p>Downloading of the Reference Datasets.</p></li>
<li><p>Quality Control and Data Preprocessing.</p></li>
<li><p>Relatedness Inference Workflow.</p></li>
</ol>
<p>The pipeline is implemented with the <a class="reference external" href="https://snakemake.github.io/">Snakemake framework</a> and can be accessible through the <cite>snakemake</cite> command.</p>
<p>All the pipeline functionality is embedded into the launcher (<cite>launcher.py</cite>), which invokes Snakemake workflows under the hood.
By default, all the commands just check the accessibility of the data and build the computational graph.</p>
<p>To actually perform computation one should add <cite>–real-run</cite> flag to all the commands.</p>
<p>Read the sections below to run <em>GRAPE</em> on your data</p>
<section id="quality-control-and-data-preprocessing">
<h2>Quality Control and Data Preprocessing<a class="headerlink" href="#quality-control-and-data-preprocessing" title="Link to this heading"></a></h2>
<p>GRAPE have a versatile and configurable preprocessing workflow.
One part of the preprocessing is required and must be performed before the relationship inference workflow.</p>
<p>Preprocessing is launched by the <cite>preprocess</cite> command of the GRAPE.</p>
<section id="preprocessing-steps">
<h3>Preprocessing steps<a class="headerlink" href="#preprocessing-steps" title="Link to this heading"></a></h3>
<p>Along with some necessary technical procedures, preprocessing includes the following steps:</p>
<ul class="simple">
<li><dl class="simple">
<dt><strong>[Required] SNPs quality control by minor allele frequency (MAF) and the missingness rate</strong>.</dt><dd><p>We discovered that blocks of rare SNPs with low MAF value in genotype arrays may produce false positive IBD segments.
To address this problem, we filter SNPs by minor allele frequency.
We remove SNPs with MAF value less than 0.02.
Additionally, we remove multiallelic SNPs, insertions / deletions, and SNPs with the high missingness rate, because such SNPs are inconsistent with IBD detection tools.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[Required] Per-sample quality control, using missingness and heterozygosity</strong>.</dt><dd><p>Extensive testing revealed that samples with an unusually low level of heterozygosity could produce many false relatives matches among individuals.
GRAPE excludes such samples from the analysis and creates a report file with the description of the exclusion reason.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[Required] Control for strands and SNP IDs mismatches</strong>.</dt><dd><p>During this step GRAPE fixes inconsistencies in strands and reference alleles.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[Optional] LiftOver from hg38 to hg37</strong>.</dt><dd><p>Currently GRAPE uses hg37 build version of the human genome reference.
The pipeline supports input in hg38 and hg37 builds (see <cite>–assembly</cite> flag of the pipeline launcher).
If hg38 build is selected (<cite>–assembly h38</cite>), then GRAPE applies liftOver tool to the input data in order to match the hg37 reference assembly.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[Optional] Phasing and imputation</strong>.</dt><dd><p>GRAPE supports phasing and genotype imputation using 1000 Genomes Project reference panel.
GERMLINE IBD detection tool requires phased data.
So, if input data is unphased, one should include phasing (Eagle 2.4.1) (<cite>–phase</cite> flag) into the preprocessing before running the GERMLINE workflow.
If input data is highly _heterogeneous_ in a sense of available SNPs positions, one can also add imputation (Minimac4) procedure to the preprocessing (<cite>–impute</cite> flag).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[Optional] Removal of imputed SNPs</strong>.</dt><dd><p>We found, that if input data is _homogeneous_ in a sense of SNPs positions, the presence of imputed SNPs does not affect the overall IBD detection accuracy of the IBIS tool, but it significantly slows down the overall performance.
For this particular case, when input data initially contains a lot of imputed SNPs, we recommend to remove them by specifying <cite>–remove-imputation</cite> flag to the GRAPE launcher.
GRAPE removes all SNPs which are marked with <cite>IMPUTED</cite> flag in the input VCF file.</p>
</dd>
</dl>
</li>
</ul>
</section>
<section id="workflows-usage">
<h3>Workflows Usage<a class="headerlink" href="#workflows-usage" title="Link to this heading"></a></h3>
<ol class="arabic simple">
<li><dl class="simple">
<dt>Preprocessing for the IBIS + KING relatedness inference workflow.</dt><dd><ul class="simple">
<li><p>Input file is located at <cite>/media/input.vcf.gz</cite>.</p></li>
<li><p>GRAPE working directory is <cite>/media/data</cite>.</p></li>
<li><p>Assembly of the input VCF file is <cite>hg37</cite>.</p></li>
</ul>
</dd>
</dl>
</li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">--</span><span class="n">rm</span> <span class="o">-</span><span class="n">it</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">media</span><span class="p">:</span><span class="o">/</span><span class="n">media</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">localtime</span><span class="p">:</span><span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">localtime</span><span class="p">:</span><span class="n">ro</span> \
    <span class="n">grape</span><span class="p">:</span><span class="n">latest</span> <span class="n">launcher</span><span class="o">.</span><span class="n">py</span> <span class="n">preprocess</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">ref</span> \
        <span class="o">--</span><span class="n">vcf</span><span class="o">-</span><span class="n">file</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="nb">input</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">--</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span> <span class="o">--</span><span class="n">assembly</span> <span class="n">hg37</span> <span class="o">--</span><span class="n">real</span><span class="o">-</span><span class="n">run</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><dl class="simple">
<dt>Preprocessing for the GERMLINE + KING workflow.</dt><dd><ul class="simple">
<li><p>Assembly of the input VCF file is <cite>hg38</cite>.</p></li>
<li><p>GERMLINE can work with phased data only, so we add phasing procedure to the preprocessing.</p></li>
<li><p>Genotype imputation is also added.</p></li>
</ul>
</dd>
</dl>
</li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">--</span><span class="n">rm</span> <span class="o">-</span><span class="n">it</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">media</span><span class="p">:</span><span class="o">/</span><span class="n">media</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">localtime</span><span class="p">:</span><span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">localtime</span><span class="p">:</span><span class="n">ro</span> \
    <span class="n">grape</span><span class="p">:</span><span class="n">latest</span> <span class="n">launcher</span><span class="o">.</span><span class="n">py</span> <span class="n">preprocess</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">ref</span> \
        <span class="o">--</span><span class="n">vcf</span><span class="o">-</span><span class="n">file</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="nb">input</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">--</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span> \
        <span class="o">--</span><span class="n">assembly</span> <span class="n">hg38</span> <span class="o">--</span><span class="n">phase</span> <span class="o">--</span><span class="n">impute</span> <span class="o">--</span><span class="n">real</span><span class="o">-</span><span class="n">run</span>
</pre></div>
</div>
</section>
</section>
<section id="relatedness-inference">
<h2>Relatedness Inference<a class="headerlink" href="#relatedness-inference" title="Link to this heading"></a></h2>
<p>There are three relationship inference workflows implemented in GRAPE.</p>
<p>These workflows are activated by the <cite>find</cite> command of the launcher.</p>
<p>Workflow selection is made by the <cite>–flow</cite> parameter.</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt><strong>IBIS + ERSA</strong>, <cite>–flow ibis</cite>.</dt><dd><p>During this workflow IBD segments detection is performed by IBIS, and estimation of relationship degree is carried out by means of ERSA algorithm.
This is the fastest workflow.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>IBIS + ERSA &amp; KING</strong>, <cite>–flow ibis-king</cite>.</dt><dd><p>During this workflow, GRAPE uses KING tool for the first 3 degrees of relationships, and IBIS + ERSA approach for higher order degrees.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>GERMLINE + ERSA &amp; KING</strong>, <cite>–flow germline-king</cite>.</dt><dd><p>The workflow uses GERMLINE for IBD segments detection.
KING is used to identify relationship for the first 3 degrees, and ERSA algorithm is used for higher order degrees.
This workflow was added to GRAPE mainly for the case, when input data is already phased and accurately preprocessed.</p>
</dd>
</dl>
</li>
</ol>
<section id="id1">
<h3>Workflows Usage<a class="headerlink" href="#id1" title="Link to this heading"></a></h3>
<ol class="arabic simple">
<li><p>Relationship inference with the IBIS + ERSA &amp; KING workflow.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">--</span><span class="n">rm</span> <span class="o">-</span><span class="n">it</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">media</span><span class="p">:</span><span class="o">/</span><span class="n">media</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">localtime</span><span class="p">:</span><span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">localtime</span><span class="p">:</span><span class="n">ro</span> \
    <span class="n">grape</span><span class="p">:</span><span class="n">latest</span> <span class="n">launcher</span><span class="o">.</span><span class="n">py</span> <span class="n">find</span> <span class="o">--</span><span class="n">flow</span> <span class="n">ibis</span><span class="o">-</span><span class="n">king</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">ref</span> \
        <span class="o">--</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span> <span class="o">--</span><span class="n">ibis</span><span class="o">-</span><span class="n">seg</span><span class="o">-</span><span class="nb">len</span> <span class="mi">7</span> <span class="o">--</span><span class="n">ibis</span><span class="o">-</span><span class="nb">min</span><span class="o">-</span><span class="n">snp</span> <span class="mi">500</span> \
        <span class="o">--</span><span class="n">zero</span><span class="o">-</span><span class="n">seg</span><span class="o">-</span><span class="n">count</span> <span class="mf">0.5</span> <span class="o">--</span><span class="n">zero</span><span class="o">-</span><span class="n">seg</span><span class="o">-</span><span class="nb">len</span> <span class="mf">5.0</span> <span class="o">--</span><span class="n">alpha</span> <span class="mf">0.01</span> <span class="o">--</span><span class="n">real</span><span class="o">-</span><span class="n">run</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p>Relationship inference with GERMLINE + ERSA &amp; KING workflow.</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">--</span><span class="n">rm</span> <span class="o">-</span><span class="n">it</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">media</span><span class="p">:</span><span class="o">/</span><span class="n">media</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">localtime</span><span class="p">:</span><span class="o">/</span><span class="n">etc</span><span class="o">/</span><span class="n">localtime</span><span class="p">:</span><span class="n">ro</span> \
    <span class="n">grape</span><span class="p">:</span><span class="n">latest</span> <span class="n">launcher</span><span class="o">.</span><span class="n">py</span> <span class="n">find</span> <span class="o">--</span><span class="n">flow</span> <span class="n">germline</span><span class="o">-</span><span class="n">king</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">ref</span> \
        <span class="o">--</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span> <span class="o">--</span><span class="n">zero</span><span class="o">-</span><span class="n">seg</span><span class="o">-</span><span class="n">count</span> <span class="mf">0.5</span> <span class="o">--</span><span class="n">zero</span><span class="o">-</span><span class="n">seg</span><span class="o">-</span><span class="nb">len</span> <span class="mf">5.0</span> \
        <span class="o">--</span><span class="n">alpha</span> <span class="mf">0.01</span> <span class="o">--</span><span class="n">real</span><span class="o">-</span><span class="n">run</span>
</pre></div>
</div>
</section>
<section id="description-of-the-ibis-and-ersa-parameters">
<h3>Description of the IBIS and ERSA Parameters<a class="headerlink" href="#description-of-the-ibis-and-ersa-parameters" title="Link to this heading"></a></h3>
<ul class="simple">
<li><dl class="simple">
<dt><strong>[IBIS]</strong> <cite>–ibis-seg-len</cite>.</dt><dd><p>Minimum length of the IBD segment to be found by IBIS.
Higher values reduce false positive rate and give less distant matches (default = 7 cM).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[IBIS]</strong> <cite>–ibis-min-snp</cite>.</dt><dd><p>Minimum number of SNPs per IBD segment to be detected (default = 500 SNPs).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[ERSA]</strong> <cite>–zero-seg-count</cite>.</dt><dd><p>Mean number of shared segments for two unrelated individuals in the population.
Smaller values tend to give more distant matches and increase the false positive rate (default = 0.5).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[ERSA]</strong> <cite>–zero-seg-len</cite>.</dt><dd><p>Average length of IBD segment for two unrelated individuals in the population.
Smaller values tend to give more distant matches and increase the false positive rate (default = 5 cM).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>[ERSA]</strong> <cite>–alpha</cite>.</dt><dd><p>ERSA significance level (default = 0.01).</p>
</dd>
</dl>
</li>
</ul>
</section>
<section id="description-of-the-output-relatives-file">
<h3>Description of the Output Relatives File<a class="headerlink" href="#description-of-the-output-relatives-file" title="Link to this heading"></a></h3>
<p>Output relatives file is in TSV file format.
It contains one line for each detected pair of relatives.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">id1</span>      <span class="n">id2</span>     <span class="n">king_degree</span> <span class="n">king_relation</span> <span class="n">shared_genome_proportion</span> <span class="n">kinship</span> <span class="n">kinship_degree</span> <span class="n">ersa_degree</span> <span class="n">ersa_lower_bound</span> <span class="n">ersa_upper_bound</span> <span class="n">shared_ancestors</span> <span class="n">final_degree</span>  <span class="n">total_seg_len</span>        <span class="n">seg_count</span>
<span class="n">g1</span><span class="o">-</span><span class="n">b1</span><span class="o">-</span><span class="n">i1</span> <span class="n">g2</span><span class="o">-</span><span class="n">b1</span><span class="o">-</span><span class="n">i1</span>     <span class="mi">1</span>           <span class="n">PO</span>                <span class="mf">0.4996</span>          <span class="mf">0.2493</span>      <span class="mf">1.0</span>             <span class="mi">1</span>               <span class="mi">1</span>               <span class="mi">1</span>               <span class="mf">0.0</span>             <span class="mi">1</span>       <span class="mf">3359.9362945470302</span>      <span class="mi">40</span>
<span class="n">g1</span><span class="o">-</span><span class="n">b1</span><span class="o">-</span><span class="n">i1</span> <span class="n">g2</span><span class="o">-</span><span class="n">b2</span><span class="o">-</span><span class="n">i1</span>     <span class="mi">1</span>           <span class="n">PO</span>                <span class="mf">0.4997</span>          <span class="mf">0.2459</span>      <span class="mf">1.0</span>             <span class="mi">1</span>               <span class="mi">1</span>               <span class="mi">1</span>               <span class="mf">0.0</span>             <span class="mi">1</span>       <span class="mf">3362.012253715002</span>       <span class="mi">40</span>
<span class="n">g1</span><span class="o">-</span><span class="n">b1</span><span class="o">-</span><span class="n">i1</span> <span class="n">g2</span><span class="o">-</span><span class="n">b3</span><span class="o">-</span><span class="n">i1</span>     <span class="mi">1</span>           <span class="n">PO</span>                <span class="mf">0.4999</span>          <span class="mf">0.2467</span>      <span class="mf">1.0</span>             <span class="mi">1</span>               <span class="mi">1</span>               <span class="mi">1</span>               <span class="mf">0.0</span>             <span class="mi">1</span>       <span class="mf">3363.150814131464</span>       <span class="mi">40</span>
<span class="n">g1</span><span class="o">-</span><span class="n">b1</span><span class="o">-</span><span class="n">i1</span> <span class="n">g3</span><span class="o">-</span><span class="n">b1</span><span class="o">-</span><span class="n">i1</span>     <span class="mi">2</span>           <span class="mi">2</span>                 <span class="mf">0.2369</span>          <span class="mf">0.1163</span>      <span class="mf">2.0</span>             <span class="mi">2</span>               <span class="mi">2</span>               <span class="mi">2</span>               <span class="mf">1.0</span>             <span class="mi">2</span>       <span class="mf">1644.634182188072</span>       <span class="mi">60</span>
</pre></div>
</div>
<dl>
<dt>Here is the description of the columns:</dt><dd><ul>
<li><p><cite>id1</cite> - ID of the first sample in a pair of relatives.</p></li>
<li><p><cite>id2</cite> - ID of the second sample in a pair of relatives, <cite>id1</cite> is always less than <cite>id2</cite>.</p></li>
<li><dl>
<dt><cite>king_degree</cite> - Numeric degree of relationship estimated by KING;</dt><dd><ul class="simple">
<li><p><cite>0</cite> means duplicates or MZ twins,</p></li>
<li><p><cite>1</cite> means parent-offspring (PO),</p></li>
<li><p><cite>2</cite> can be either full siblings (FS), half siblings and grandmother/grandfather with a granddaughter/grandson.</p></li>
<li><p><cite>3</cite> is aunt/uncle with a niece/nephew, as described in table in <a class="reference external" href="https://en.wikipedia.org/wiki/Coefficient_of_relationship">https://en.wikipedia.org/wiki/Coefficient_of_relationship</a>.</p></li>
</ul>
<p>If <cite>king_degree</cite> exists, then <cite>final_degree</cite> will be equal to <cite>king_degree</cite>.</p>
</dd>
</dl>
</li>
<li><p><cite>king_relation</cite> - further differentiation for the first 2 degrees of KING; <cite>king_degree</cite> equals <cite>1</cite> means PO - parent-offspring, also KING detects FS in some of second degrees.</p></li>
<li><dl class="simple">
<dt><cite>shared_genome_proportion</cite> is the approximate fraction of genome shared by two individuals.</dt><dd><p>It should be approximately 0.5 for PO and FS, 0.25 for grandmother-granddaughter and half-siblings.
For the first 3 degrees it is calculated as total length of IBD2 segments + half of total length of IBD1 segments.
For 4th+ degrees it is simply half of total length of IBD1 segments.</p>
</dd>
</dl>
</li>
<li><p><cite>kinship</cite> is the KING kinship coefficient.</p></li>
<li><p><cite>ersa_degree</cite> is the degree estimated from IBD segments by ERSA, it is used for the <cite>final_degree</cite> when <cite>king_degree</cite> does not exist.</p></li>
<li><dl class="simple">
<dt><cite>ersa_lower_bound</cite> / <cite>ersa_upper_bound</cite> is the lower / upper bound of the degree estimated by the ERSA using confidence interval corresponding to the significance level &amp;alpha; (<cite>–alpha</cite>).</dt><dd><p>Accordingly to the ERSA likelihood model, true degree does not belong to this interval with probability equals to &amp;alpha;.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>shared_ancestors</cite> is the most likeliest number of shared ancestors.</dt><dd><p>If it equals <cite>0</cite>, then one relative is a direct descendant of the other; if equals <cite>1</cite>, then they most probably have one common ancestor and represent half siblings; if equals <cite>2</cite>, then, for examples, individuals may have common mother and father.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>final_degree</cite> is the final relationship degree predicted by GRAPE.</dt><dd><p>If KING is used (<cite>–flow ibis-king</cite> or <cite>–flow germline-king</cite>), it equals <cite>king_degree</cite> for close relatives up to the 3rd degree, and <cite>ersa_degree</cite> for distant relatives.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>total_seg_len</cite> is the total length of all IBD1 segments.</dt><dd><p>It’s calculated using IBIS or GERMLINE IBD data.
If KING is involved, then for the first 3 degrees it’s calculated using KING IBD data.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>seg_count</cite> is the total number of all IBD segments found by IBIS / GERMLINE tools.</dt><dd><p>If KING is involved, the total number of found IBD segments is taken from KING for the first 3 degrees.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</section>
</section>
<section id="ibd-segments-weighting">
<h2>IBD Segments Weighting<a class="headerlink" href="#ibd-segments-weighting" title="Link to this heading"></a></h2>
<p>Distribution of IBD segments among non-related (background) individuals within a population may be quite heterogeneous.
There may exist genome regions with extremely high rates of overall matching, which are not inherited from the recent common ancestors.
Instead, these regions more likely reflect other demographic factors of the population.</p>
<p>The implication is that IBD segments detected in such regions are expected to be less useful for estimating recent relationships.
Moreover, such regions potentially prone to false-positive IBD segments.</p>
<p>GRAPE provides two options to address this issue:</p>
<section id="genome-regions-exclusion-mask">
<h3>Genome Regions Exclusion Mask<a class="headerlink" href="#genome-regions-exclusion-mask" title="Link to this heading"></a></h3>
<p>Some regions are completely excluded from the consideration.
This approach is implemented in ERSA and is used by GRAPE by default.</p>
</section>
<section id="ibd-segments-weighing">
<h3>IBD Segments Weighing<a class="headerlink" href="#ibd-segments-weighing" title="Link to this heading"></a></h3>
<p>The key idea is to down-weight IBD segment, i.e. reduce the IBD segment length, if the segment cross regions with high rate of matching.
Down-weighted segments are then passed to the ERSA algorithm.</p>
<p>GRAPE provides an ability to compute the weight mask from the VCF file with presumably unrelated individuals.</p>
<p>This mask is used during the relatedness detection by specifying the <cite>–weight-mask</cite> flag of the launcher.</p>
<p>See more information in the <a class="reference external" href="https://f1000research.com/articles/11-589">GRAPE paper</a>.</p>
</section>
<section id="computation-of-the-ibd-segments-weighing-mask">
<h3>Computation of the IBD segments weighing mask<a class="headerlink" href="#computation-of-the-ibd-segments-weighing-mask" title="Link to this heading"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">--</span><span class="n">rm</span> <span class="o">-</span><span class="n">it</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">media</span><span class="p">:</span><span class="o">/</span><span class="n">media</span> \
    <span class="n">grape</span><span class="p">:</span><span class="n">latest</span> <span class="n">launcher</span><span class="o">.</span><span class="n">py</span> <span class="n">compute</span><span class="o">-</span><span class="n">weight</span><span class="o">-</span><span class="n">mask</span> \
        <span class="o">--</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">background</span> <span class="o">--</span><span class="n">assembly</span> <span class="n">hg37</span> \
        <span class="o">--</span><span class="n">real</span><span class="o">-</span><span class="n">run</span> <span class="o">--</span><span class="n">ibis</span><span class="o">-</span><span class="n">seg</span><span class="o">-</span><span class="nb">len</span> <span class="mi">5</span> <span class="o">--</span><span class="n">ibis</span><span class="o">-</span><span class="nb">min</span><span class="o">-</span><span class="n">snp</span> <span class="mi">400</span>
</pre></div>
</div>
<p>The resulting files consist of a weight mask file in JSON format and a visualization of the mask stored in <cite>/media/background/weight-mask/</cite> folder.</p>
</section>
<section id="usage-of-the-ibd-segments-weighing-mask">
<h3>Usage of the IBD segments weighing mask<a class="headerlink" href="#usage-of-the-ibd-segments-weighing-mask" title="Link to this heading"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">--</span><span class="n">rm</span> <span class="o">-</span><span class="n">it</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">media</span><span class="p">:</span><span class="o">/</span><span class="n">media</span> \
    <span class="n">grape</span><span class="p">:</span><span class="n">latest</span> <span class="n">launcher</span><span class="o">.</span><span class="n">py</span> <span class="n">find</span> <span class="o">--</span><span class="n">flow</span> <span class="n">ibis</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">ref</span> \
        <span class="o">--</span><span class="n">weight</span><span class="o">-</span><span class="n">mask</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">background</span><span class="o">/</span><span class="n">weight</span><span class="o">-</span><span class="n">mask</span><span class="o">/</span><span class="n">mask</span><span class="o">.</span><span class="n">json</span> \
        <span class="o">--</span><span class="n">directory</span> <span class="o">/</span><span class="n">media</span><span class="o">/</span><span class="n">data</span> <span class="o">--</span><span class="n">assembly</span> <span class="n">hg37</span> \
        <span class="o">--</span><span class="n">real</span><span class="o">-</span><span class="n">run</span> <span class="o">--</span><span class="n">ibis</span><span class="o">-</span><span class="n">seg</span><span class="o">-</span><span class="nb">len</span> <span class="mi">5</span> <span class="o">--</span><span class="n">ibis</span><span class="o">-</span><span class="nb">min</span><span class="o">-</span><span class="n">snp</span> <span class="mi">400</span>
</pre></div>
</div>
</section>
</section>
<section id="execution-by-scheduler">
<h2>Execution by Scheduler<a class="headerlink" href="#execution-by-scheduler" title="Link to this heading"></a></h2>
<p>The pipeline can be executed using lightweight scheduler <a class="reference external" href="https://ohsu-comp-bio.github.io/funnel">Funnel</a>,
which implements <a class="reference external" href="https://github.com/ga4gh/task-execution-schemas">Task Execution Schema</a> developed by <a class="reference external" href="https://github.com/ga4gh/wiki/wiki">GA4GH</a>.</p>
<p>During execution, incoming data for analysis can be obtained in several ways: locally, FTP, HTTPS, S3, Google, etc.
The resulting files can be uploaded in the same ways.</p>
<p>It is possible to add other features such as writing to the database, and sending to the REST service.</p>
<p>The scheduler itself can work in various environments from a regular VM to a Kubernetes cluster with resource quotas support.</p>
<p>For more information see the <a class="reference external" href="https://ohsu-comp-bio.github.io/funnel/docs">Funnel Documentation</a>.</p>
<p>How to use Funnel:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Start the Funnel server</span>
<span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">funnel</span> <span class="n">server</span> <span class="n">run</span>

<span class="c1"># Use Funnel as client</span>
<span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">funnel</span> <span class="n">task</span> <span class="n">create</span> <span class="n">funnel</span><span class="o">/</span><span class="n">grape</span><span class="o">-</span><span class="n">real</span><span class="o">-</span><span class="n">run</span><span class="o">.</span><span class="n">json</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="dockstore.html" class="btn btn-neutral float-right" title="Dockstore" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2023, GENXT.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>